src/ucp/dt/dt_generic.c

```diff
@@ -27,9 +27,11 @@ ucs_status_t ucp_dt_create_generic(const ucp_generic_dt_ops_t *ops, void *contex
         return UCS_ERR_NO_MEMORY;
     }
 
-    dt_gen->ops     = *ops;
-    dt_gen->context = context;
-    *datatype_p     = ucp_dt_from_generic(dt_gen);
+    dt_gen->ops         = *ops;
+    dt_gen->context     = context;
+    dt_gen->packed_size = 0;
```

Comment:
I actually used 0 to allow checks like `if (attr->packed_size)`, but I changed it according to your suggestion, so we can also use `SIZE_MAX` now. I think that a datatype of size 0 does not make sense. If you still think it's better to initialize with `SIZE_MAX`, I am okay doing so, especially since I probably don't know all the data descriptions/types that are possible with UCX.

Reply:
i think having 0-sized type is more likely than SIZE_MAX, so I'd change it.
@yosefe, wdyt?

Reply:
I don't think we can assume that packed size of bufferA is same as packed size of bufferB. It depends on the contents.
Perhaps dt_query routine should also accept buffer pointer and count.

Reply:
You are right, if we consider compression, the packed size of a generic datatype can be different depending on the data. I thought of packing more like C struct packing/padding.
This basically means that the packed size can be different for each call to a communication function. Hence, the query would always return the same value for contiguous datatypes and could return different values after each communication operation for the same generic datatype. The function then still serves its purpose and allows me to find out how much data is transferred.

Reply:
If we add the buffer pointer and count to `dt_query`, we could also support IOV. Should we then return the packed message/buffer size instead of the packed datatype size? Alternatively, we would have to divide by count.
I assume that `start_pack` and `packed_size` are short calls, so that is doesn't hurt calling them twice (a second time for the size query).

Reply:
Has a decision been made here? I agree with @yosefe that passing buffer and count and return the buffer size fits better to the existing UCP datatypes. The user of `ucp_dt_query` can still pass "1" as count to return the size of one element.

Reply:
@shamisp ping on this

Reply:
@yosefe This is good question actually. In our recent research work the size also depended on the target EP in addition to buffer and count - think about DT conversion across different architectures. Do we use this for any of the MPI call ? I don't think MPI passes buffer (I have to double check)

Reply:
I did not come across the EP during implementation. The current implementation basically returns the last datatype size, which has been determined during the regular message/transfer preparation. This should include the EP.